src/pg/decoder.cr

```diff
@@ -294,12 +317,56 @@ module PG
       end
     end
 
-    @@decoders = Hash(Int32, PG::Decoders::Decoder).new(ByteaDecoder.new)
+    alias DecoderMap = Hash(Int32, PG::Decoders::Decoder)
+    @@decoders = DecoderMap.new(ByteaDecoder.new)
 
     def self.from_oid(oid)
       @@decoders[oid]
     end
 
+    TYPE_SQL = %q(
+      SELECT oid, typname, typcategory
+      FROM pg_type
+      WHERE typisdefined = 't'
+        AND typtype IN ('b', 'd'))
+
+    # https://www.postgresql.org/docs/9.4/static/catalog-pg-type.html#CATALOG-TYPCATEGORY-TABLE
+    module TypeCategories # Could this be an enum?
+      ARRAY        = 'A'
+      COMPOSITE    = 'C'
+      DATE_OR_TIME = 'D'
+      ENUM         = 'E'
+      GEOMETRIC    = 'G'
+      NETWORK      = 'I'
+      NUMERIC      = 'N'
+      PSEUDO       = 'P'
+      RANGE        = 'R'
+      STRING       = 'S'
+      TIMESPAN     = 'T'
+      USER_DEFINED = 'U'
+      BIT_STRING   = 'V'
+      UNKNOWN      = 'X'
+    end
+
+    # Builds a `DecoderMap` of all types visible on `connection` that are not
+    # already statically known.
+    def self.register_connection_decoders(connection : PG::Connection) : Void
+      types = connection.query_all(TYPE_SQL, as: {UInt32, String, Char})
+      types.each do |oid, name, category|
+        oid = oid.to_i32 # Query execution if I read straight to Int32 :\
```

Comment:
Oops this comment should read "query execution hangs if...".

That was a reminder to me to ask @will why that is? I was fighting a lot of mysterious hangs if I didn't get things right. It can be hard to debug sometimes. Any advice?